### PR TITLE
merge arktos-up-scale-out.sh from poc branch

### PR DESCRIPTION
**Pull request by pdgetrf**
## Changes
Merged arktos-up-scale-out.sh related changes from poc.

## Notes to reviewers
- arktos-up-scale-out.sh is added from poc without any change.
- further refactor possibly needed to combine arktos-up-scale-out.sh with arktos-up.sh

## Validation
- arktos-up.sh runs correctly
- arktos-up-scale-out.sh runs correctly

"correctly" here means it's verified with pod & deployments creation and deletion (2tp/1rp with a tenant from both tp for the case of scale-out)

### CODE REVIEW

File: hack/lib/common.sh [hunk 1]
```diff
@@ -474,6 +474,19 @@ function kube::common::start_controller_manager {
       cloud_config_arg+=("--external-cloud-volume-plugin=${CLOUD_PROVIDER}")
       cloud_config_arg+=("--cloud-config=${CLOUD_CONFIG}")
     fi
+
+    if [[ -z "${IS_RESOURCE_PARTITION}" ]]; then
+       MASTER_ENDPOINT="https://${API_HOST}:${API_SECURE_PORT}"
```
Comment (Sindica) [inline hack/lib/common.sh:479]: Perhaps MASTER_ENDPOINT only need to be set once in common.sh?

**Comment by Sindica** [comment]
Somehow arktos-up.sh cannot find node on my dev box. Did you try both scale-up and arktos-up on the same machine?

**Comment by pdgetrf** [comment]
> Somehow arktos-up.sh cannot find node on my dev box. Did you try both scale-up and arktos-up on the same machine?

on different machines

**Comment by pdgetrf** [comment]
> Somehow arktos-up.sh cannot find node on my dev box. Did you try both scale-up and arktos-up on the same machine?

tried again on a separate machine. works fine.

image

**Comment by Sindica** [comment]
> > Somehow arktos-up.sh cannot find node on my dev box. Did you try both scale-up and arktos-up on the same machine?
> 
> tried again on a separate machine. works fine.
> 
> image

My bad, looks like I have env setup for scale out when I tried to use arktos-up. One second thought, can you add env clean script to arktos-up.sh to saving dev from trouble shooting?

unset SCALE_OUT_PROXY_ENDPOINT
unset TENANT_PARTITION_IP
unset RESOURCE_PARTITION_IP
unset IS_RESOURCE_PARTITION
unset TENANT_SERVERS

Consider add a manual for hack/arktos-up-scale-out-poc.sh

**Review by Sindica — COMMENTED** [review]
Running arktos-up-scale-out-poc.sh on RP, KCM exit unexpectedly. Got following error:
I0222 20:47:50.692806    4190 controllermanager.go:534] Starting "nodelifecycle"
E0222 20:47:50.693163    4190 controllermanager.go:537] Error starting "nodelifecycle"
F0222 20:47:50.693176    4190 controllermanager.go:241] error starting controllers: Invalid url (http://172.30.0.148): invalid host IP

**Comment by pdgetrf** [comment]
> Running arktos-up-scale-out-poc.sh on RP, KCM exit unexpectedly. Got following error:
> I0222 20:47:50.692806 4190 controllermanager.go:534] Starting "nodelifecycle"
> E0222 20:47:50.693163 4190 controllermanager.go:537] Error starting "nodelifecycle"
> F0222 20:47:50.693176 4190 controllermanager.go:241] error starting controllers: Invalid url (http://172.30.0.148): invalid host IP

I tested it this morning after another rebase. Didn't see this problem.

**Comment by Sindica** [comment]
Xiaoning and me had a quick discussion today. This is going to master branch. Will like to have a simple instruction on how to use this script, such as required environment setup. It will be even better if there can be initial validation and instruction output if validation is failed.

**Comment by pdgetrf** [comment]
> Xiaoning and me had a quick discussion today. This is going to master branch. Will like to have a simple instruction on how to use this script, such as required environment setup. It will be even better if there can be initial validation and instruction output if validation is failed.
 
We could either discard this PR or merge it with fixes, instead of planning on new features that are not even implemented yet. Keeping adding new features to a PR in review is not the right process to treat PR.

File: hack/arktos-up-scale-out-poc.sh [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+echo KUBELET_FLAGS for new kubelet commandline --tenant-servers
+echo ${KUBELET_FLAGS}
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+
+if [[ -z "${TENANT_SERVERS}" ]]; then
+  TENANT_SERVERS=${SCALE_OUT_PROXY_ENDPOINT}
+fi
+
+# for POC, the kubelet_flags is used for the new temporary kubelet commandline args
```
Comment (Sindica) [inline hack/arktos-up-scale-out-poc.sh:38]: Line 26-44 has duplicated logic.

Comment (pdgetrf) [reply hack/arktos-up-scale-out-poc.sh:38]: removed.

File: hack/arktos-up-scale-out-poc.sh [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+  fi
+fi
+
+
+
+# Ensure CERT_DIR is created for auto-generated crt/key and kubeconfig
+mkdir -p "${CERT_DIR}" &>/dev/null || sudo mkdir -p "${CERT_DIR}"
+CONTROLPLANE_SUDO=$(test -w "${CERT_DIR}" || echo "sudo -E")
+
+cleanup()
```
Comment (Sindica) [inline hack/arktos-up-scale-out-poc.sh:165]: function cleanup, healthcheck, print_color, warning_log, start_kubedns, start_nodelocaldns, start_kubedashboard, create_psp_policy, create_storage_class are same in arktos-up-scale-out-poc.sh and arktos-up.sh
Can they be refactored?

File: hack/arktos-up-scale-out-poc.sh [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+  echo -n "$(tput bold)$(tput setaf "${color}")"
+  echo "${prefix}${message}"
+  echo -n "$(tput sgr0)"
+}
+
+function warning_log {
+  print_color "$1" "W$(date "+%m%d %H:%M:%S")]" 1
+}
+
+function start_etcd {
```
Comment (Sindica) [inline hack/arktos-up-scale-out-poc.sh:280]: Also start_etcd has same implementation, not sure whether this can be refactored

**Comment by Sindica** [comment]
> > Xiaoning and me had a quick discussion today. This is going to master branch. Will like to have a simple instruction on how to use this script, such as required environment setup. It will be even better if there can be initial validation and instruction output if validation is failed.
> 
> We could either discard this PR or merge it with fixes, instead of planning on new features that are not even implemented yet. Keeping adding new features to a PR in review is not the right process to treat PR.

Peng, I have many trouble testing this script. Please provide some instruction about how to set up environment either in the script or as a separate manual. Otherwise, it will be hard for others to use this script.

File: hack/arktos-up-scale-out-poc.sh (hunk 4 above)
Comment (pdgetrf) [reply hack/arktos-up-scale-out-poc.sh:280]: As stated in "Notes to reviewers" section in the description, refactor would be better for a following PR.

Comment (pdgetrf) [reply hack/arktos-up-scale-out-poc.sh:280]: I will draft a separate doc on usage

**Comment by pdgetrf** [comment]
> > > Xiaoning and me had a quick discussion today. This is going to master branch. Will like to have a simple instruction on how to use this script, such as required environment setup. It will be even better if there can be initial validation and instruction output if validation is failed.
> > 
> > 
> > We could either discard this PR or merge it with fixes, instead of planning on new features that are not even implemented yet. Keeping adding new features to a PR in review is not the right process to treat PR.
> 
> Peng, I have many trouble testing this script. Please provide some instruction about how to set up environment either in the script or as a separate manual. Otherwise, it will be hard for others to use this script.

sure. I will draft a separate doc on usage

File: hack/arktos-up-scale-out-poc.sh (hunk 3 above)
Comment (pdgetrf) [reply hack/arktos-up-scale-out-poc.sh:165]: please see answer below

File: hack/lib/common.sh (hunk 1 above)
Comment (pdgetrf) [reply hack/lib/common.sh]: this is refactored.

File: hack/arktos-up-scale-out-poc.sh (hunk 4 above)
Comment (Sindica) [reply hack/arktos-up-scale-out-poc.sh:276]: What is the benefit of refactor later? I only see allowing others to start using the code before refactor is done. We don't have the pressing need for now.

Comment (pdgetrf) [reply hack/arktos-up-scale-out-poc.sh:276]: It goes beyond the purpose of this PR and I don't think keeping adding tasks to an existing scope of a PR is the right approach. Implementing a function and then refactor is a common practice to keep dev cycle short and prevent regression. If there's no pressing need for now, I'd suggest to discard this PR and have a dedicated task to merge and refactor the scripts by the scalability team.

**Pull request closed without merging**